### ISSUE TITLE
Loading models to CPU in different threads

### ISSUE THREAD

**Issue by anmikh** [openvinotoolkit/openvino#487]
Is it possible to load networks in different threads via Core::LoadNetwork? 

This function seems to be not thread-safe and thus mutex is required. It leads to sequential loading models in different threads and to increasing total time of loading models as a result.

Code that reproduces segfaults / terminates of the program without using mutex:

```C++
#include <string>
#include <thread>
#include <inference_engine.hpp>

using namespace InferenceEngine;

#define WEIGHTS_EXT ".bin"
#define DEVICE_CPU "CPU"

void LoadModel(const std::string& input_model) {
    Core ie;
    CNNNetwork network = ie.ReadNetwork(input_model,
            input_model.substr(0, input_model.size() - 4) + WEIGHTS_EXT);
    network.setBatchSize(1);
    std::cout << "Thread " << std::this_thread::get_id()
              << ": Start loading model" << std::endl;
    ExecutableNetwork executable_network = ie.LoadNetwork(network, DEVICE_CPU);
    std::cout << "Thread " << std::this_thread::get_id()
              << ": End loading model" << std::endl;
}

int main(int argc, char *argv[]) {
    try {
        if (argc != 3) {
            return EXIT_FAILURE;
        }
        const std::string input_model{argv[1]};
        const size_t threads_count = std::atoi(argv[2]);

        std::cout << "Start loading models in " << threads_count
                  << " different threads" << std::endl;

        std::vector<std::thread> threads(threads_count);
        for(auto & thread : threads)
            thread = std::thread(LoadModel, input_model);

        for (auto & thread : threads) {
            if (thread.joinable())
                thread.join();
        }
    } catch (const std::exception & ex) {
        std::cerr << ex.what() << std::endl;
        return EXIT_FAILURE;
    }
    std::cout << "App finished" << std::endl;
    return EXIT_SUCCESS;
}
```

The problem is reproduced on the latest OpenVino 2020.2 release for models both in v7 and v10 formats, for example, [face-detection-retail-0005](https://download.01.org/opencv/2020/openvinotoolkit/2020.2/open_model_zoo/models_bin/2/face-detection-retail-0005/FP32/) and 8 threads were used as parameters.  For OpenVino 2020.1 release, this issue seems to be not reproduced for models in old v7/v6 formats but it definitely occurs for models in new v10 format.

Verified for different cpus: 
Core(TM) i7-8700 CPU @ 3.20GHz
Xeon(R) CPU E5-2697A v4 @ 2.60GHz
Xeon(R) Gold 6150 CPU @ 2.70GHz

**Comment by ilya-lavrenov** [comment]
@anmikh could you please try this with the latest master? we have added a mutex for critical sections inside `Core` and at least for CPU plugin we are able to `LoadNetwork` in parallel threads.

**Comment by anmikh** [comment]
@ilya-lavrenov Yes, loading networks can be done in different threads on the latest master now (without using mutex in the client's code). But it seems that almost all process of loading network is locked with the mutex, so I suppose performance gain could be better 
 in case of locking with the mutex only small part of code for real critical sections where mutex is requiered.

**Comment by ilya-lavrenov** [comment]
@anmikh from my point of view we already do in a way you suggested https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/inference_engine/ie_core.cpp#L517 We lock plugins map when someone wants to access a particular plugin. After the operation is performance like LoadNetwork, QueryNetwork and so on.

Do you see actual performance degradation? Or e.g. networks are loaded sequentially? Your help will be very important regarding this question.

**Comment by anmikh** [comment]
Thanks for the response! 
I expected more performance gain when loading networks asynchronously (it's about 10-20% faster in comparison with sync loading). For example, for public model [face-detection-retail-0005](https://download.01.org/opencv/2020/openvinotoolkit/2020.2/open_model_zoo/models_bin/2/face-detection-retail-0005/FP32/) and Core(TM) i7-8700 CPU @ 3.20GHz (the min value of loading N_networks for 100 iterations is presented in the table below):
N_networks | Async, ms | Sync, ms
-- | -- | --
2 | 320 | 354
4 | 562 | 685
8 | 1095 | 1335

Nevertheless networks are loaded asynchronously, and there is no any segfaults while loading at this mode, so I think the issue can be closed.

**Comment by ilya-lavrenov** [comment]
@anmikh how mane threads are used?

**Comment by anmikh** [comment]
@ilya-lavrenov Number of threads in async mode are equal to N_networks in appropriate row of the table above

**Comment by ilya-lavrenov** [comment]
@anmikh in this case I also expect that at least in case of 2 threads we can have 2x speed-up. 

@AlexPeskov do you know if some global locks are in the CPU plugin itself?

**Comment by anmikh** [comment]
@AlexPeskov I can confirm, that additional threads are spawned internally while loading 2 networks even in sync mode (nlwp=9 on my machine) and in async mode (nlwp=17), the cpu utilization seems to be similar in both cases.

**Comment by sahira-rizvi** [comment]
Closing, feel free to re-open or start a new issues if additional assistance is needed.

Sincerely,
Sahira
